src/persist-client/src/cache.rs

```diff
@@ -312,21 +382,117 @@ enum StateCacheInit {
     NeedInit(Arc<OnceCell<Weak<dyn DynState>>>),
 }
 
+struct PubSubSubscriptionToken {
+    shard_id: ShardId,
+    metrics: Arc<Metrics>,
+    pubsub_sender: Option<Arc<dyn PubSubSender>>,
+}
+
+impl Drop for PubSubSubscriptionToken {
+    fn drop(&mut self) {
+        if let Some(pubsub_sender) = self.pubsub_sender.as_ref() {
+            pubsub_sender.unsubscribe(&self.shard_id);
+        }
+    }
+}
+
 impl StateCache {
     /// Returns a new StateCache.
-    pub fn new(metrics: Arc<Metrics>) -> Self {
-        StateCache {
+    pub fn new(
+        cfg: &PersistConfig,
+        metrics: Arc<Metrics>,
+        pubsub: Option<(Arc<dyn PubSubSender>, Box<dyn PubSubReceiver>)>,
+    ) -> Arc<Self> {
+        let (pubsub_sender, pubsub_receiver) = match pubsub {
+            None => (None, None),
+            Some((pubsub_sender, pubsub_receiver)) => (Some(pubsub_sender), Some(pubsub_receiver)),
+        };
+
+        let state_cache = Arc::new(StateCache {
+            cfg: Arc::new(cfg.clone()),
             metrics,
             states: Default::default(),
+            pubsub_sender,
+        });
+
+        let cache = Arc::clone(&state_cache);
+        if let Some(mut pubsub_receiver) = pubsub_receiver {
+            let _ = mz_ore::task::spawn(|| "persist::rpc::push_responses", async move {
+                cache.metrics.pubsub_client.receiver.connected.set(1);
+                while let Some(res) = pubsub_receiver.next().await {
+                    let timestamp = u128::from_le_bytes(
+                        <[u8; 16]>::try_from(res.timestamp.as_slice()).expect("WIP"),
+                    );
+                    match res.message {
+                        Some(proto_pub_sub_message::Message::PushDiff(diff)) => {
+                            cache.metrics.pubsub_client.receiver.push_received.inc();
+                            let shard_id = diff.shard_id.into_rust().expect("WIP");
+                            let diff = VersionedData {
+                                seqno: diff.seqno.into_rust().expect("WIP"),
+                                data: diff.diff,
+                            };
+                            debug!(
+                                "client got diff {} {} {}",
+                                shard_id,
+                                diff.seqno,
+                                diff.data.len()
+                            );
+                            cache.apply_diff(&shard_id, diff);
+                            let now = SystemTime::now()
+                                .duration_since(SystemTime::UNIX_EPOCH)
+                                .expect("failed to get millis since epoch")
+                                .as_micros();
+                            let latency = now.saturating_sub(timestamp) as f64;
```

Comment:
i think instead we should convert the message timestamp into a Duration and do the subtraction in "Duration-space"